### PR TITLE
PHPC-1849: Handle deprecation of Serializable in PHP 8.1

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/PHPC-1849

### CODE REVIEW

File: php_phongo.h
Comment (jmikola) [inline php_phongo.h:185]: Since `__serialize` will also return a newly allocated HashTable, this is no longer just used by the `get_debug_info` handlers. `is_temp` seemed like a more accurate name.

File: src/BSON/Binary.c
Comment (jmikola) [inline src/BSON/Binary.c:308]: These protos are incorrect. The methods should have a `void` return annotation and take an `array $data` parameter and . I'll revise in a subsequent commit.

File: src/BSON/Binary.c
Comment (jmikola) [inline src/BSON/Binary.c:380]: Various `get_properties_hash` functions needed to be moved earlier in the file because they're used by `__serialize` methods. The alternative was to just add forward declarations at the top of the file, but that will ultimately be sorted by [PHPC-1728](https://jira.mongodb.org/browse/PHPC-1728).

File: src/MongoDB/CursorId.c
Comment (jmikola) [inline src/MongoDB/CursorId.c:61]: The existing implementation of `Serializable::serialize()` always used `ADD_ASSOC_INT64_AS_STRING`. That didn't seem necessary if the value falls within the range of a 32-bit signed integer. For example, `WriteConcern::serialize()` encodes `wtimeout` as either an integer or string based on its range.

However, `php_phongo_cursorid_init_from_hash` currently only expects a string, so I'll either need to change that to also accept an integer or remove the conditional encoding for `is_serialize` below.

File: src/MongoDB/CursorId.c
Comment (jmikola) [inline src/MongoDB/CursorId.c:81]: Per the comment above, this will need to be removed unless `php_phongo_cursorid_init_from_hash` is improved to also accept an integer.

Since this issue wasn't caught by any new tests, I'll be sure to add a regression test before modifying either function.

File: src/MongoDB/WriteConcern.c
Comment (jmikola) [inline src/MongoDB/WriteConcern.c:301]: Before introducing `is_serialize`, `WriteConcern::serialize()` would enter the `! is_bson` code path and conditionally encode `wtimeout` based on the system architecture. That actually seemed like a mistake, since it could lead to a 64-bit system producing serialized output that cannot be processed on a 32-bit system (e.g. if `wtimeout` was `INT32_MAX + 1`).

This adds a new code path to ignore the system architecture and instead encode according to the range of the value itself to maximum compatibility. The `init_from_hash` function already supports both formats, so this isn't the same issue that came up with CursorId earlier.

File: src/MongoDB/WriteConcern.c
Comment (jmikola) [inline src/MongoDB/WriteConcern.c:352]: I'm not sure if it's already covered in the tests, but I'll double-check that we're testing this code path of encoding `wtimeout` according to its range. I think some of the existing tests might be expecting either a string or integer (with `%r`), which is no longer necessary since the serialization output will now be determined entirely by the value.

File: src/MongoDB/CursorId.c
Comment (jmikola) [reply src/MongoDB/CursorId.c]: Decided to keep the old behavior of always serializing "id" as a string. Also added a regression test case to each of `cursorid-serialization-001.phpt` and `cursorid-serialization-002.phpt` for that behavior.

File: src/MongoDB/CursorId.c
Comment (jmikola) [reply src/MongoDB/CursorId.c:61]: Opted to remove conditional encoding for `is_serialize` since that always maintains existing behavior before this PR. See my comment below.

File: src/MongoDB/WriteConcern.c
Comment (jmikola) [reply src/MongoDB/WriteConcern.c:352]: Confirmed that this _is_ covered by the "64-bit wtimeout is always encoded as as string" test cases in `writeconcern-serialization-001.phpt` and `writeconcern-serialization-002.phpt`.

File: tests/bson/bson-utcdatetime-serialization_error-004.phpt
Comment (jmikola) [inline tests/bson/bson-utcdatetime-serialization_error-004.phpt:15]: This TODO was copied from `bson-utcdatetime-serialization_error-002.phpt`. I created [PHPC-1957](https://jira.mongodb.org/browse/PHPC-1957) to track this.

File: src/BSON/Binary.c
Comment (alcaeus) [inline src/BSON/Binary.c:93]: I've wondered about this before. What's the practical reason for nesting? Is it to only conditionally allocate the two `zval` structs on the stack and avoid doing so if there's no data to handle?

File: src/BSON/Binary.c
Comment (alcaeus) [inline src/BSON/Binary.c:334]: I've been wondering about this for a while as well: shouldn't this be in PHP's realm to provide arginfo structs for its own interface? We're duplicating this in a lot of our classes, where using `ai_Serializable___unserialize` would work just fine everywhere else 🤔

// cc @nikic

File: src/BSON/DBPointer.c
Comment (alcaeus) [inline src/BSON/DBPointer.c:90]: No need to change anything here, but getting to the second class in the code review I thought "I suppose this is now 20 times the same change", and I immediately catch the different code organisation between this and the Binary class up there. No need to change since I'd like to refactor this code in general: macros that call the combination of `ZVAL_*` and `zend_hash_str_update` would help since they could handle the scoping (if it's necessary, see earlier comment).

File: src/BSON/DBPointer.c
Comment (alcaeus) [inline src/BSON/DBPointer.c:235]: So, this is another one where I feel PHP could offer a macro:
```c
/* {{{ proto void MongoDB\Driver\DBPointer::__unserialize(array $data)
*/
IMPL_SERIALIZABLE___UNSERIALIZE_START()
	php_phongo_dbpointer_init_from_hash(Z_DBPOINTER_OBJ_P(getThis()), Z_ARRVAL_P(data));
IMPL_SERIALIZABLE___UNSERIALIZE_END() /* }}} */
```

Since the other logic is always the same, this may be an improvement to help reducing the menial task of adding this to multiple classes...

File: src/MongoDB/WriteConcern.c
Comment (alcaeus) [reply src/MongoDB/WriteConcern.c:301]: For what it's worth, looking at our platform stats, I think we'll stop considering 32-bit systems soon.

File: tests/bson/bson-binary-serialization_error-004.phpt
Comment (alcaeus) [inline tests/bson/bson-binary-serialization_error-004.phpt:13]: I know many of these files don't do this yet, but we should be using the `::class` constant. No need to change this in these files as they're most likely copypasta from the originals, so I'll see if I can do a nice find/replace to take care of these.

**Review by alcaeus — APPROVED** [review]
Changes look good, just a couple of questions on the code to judge potential refactoring effort.

File: src/BSON/Binary.c
Comment (jmikola) [reply src/BSON/Binary.c:93]: These aren't even conditionally allocated. We're just using braces to create a scope and allow us to declare variables inline. Our C compiler options would otherwise flag this as a warning and require that we declare these at the top of the function scope.

File: src/BSON/Binary.c
Comment (jmikola) [reply src/BSON/Binary.c:334]: Interesting point, although I could see this being problematic if PHP happened to change its arginfo between versions. I think this may relate to why interfaces in 8.1 started using _tentative_ return types, to defer the actual changes to PHP 9.0.

Allowing extensions to define their own arginfo presumably lets us use a narrower API if we want, which would still conform with the interface in core.

Comment (nikic) [reply src/BSON/Binary.c:334]: Nowadays we auto-generate all arginfo (as well as class declarations in general) anyway, so we don't have any particular interest in making reuse simpler.

File: src/BSON/DBPointer.c
Comment (jmikola) [reply src/BSON/DBPointer.c:90]: Created [PHPC-1962](https://jira.mongodb.org/browse/PHPC-1962) to track this.

File: tests/bson/bson-binary-serialization_error-004.phpt
Comment (jmikola) [reply tests/bson/bson-binary-serialization_error-004.phpt:13]: It would likely require a `sprintf` since we'd also need to capture the string length in the preceding integer. I'll leave this to you to make a ticket if you'd like to track the work.

**Comment by jmikola** [comment]
Noted one test failure on AppVeyor for 32-bit CursorId `var_dump()` output. Fixed in the most recent push and will wait for green CI before merging.

**Pull request merged**
